Sources/SemanticIndex/UpdateIndexStoreTaskDescription.swift

```diff
@@ -264,72 +276,150 @@ package struct UpdateIndexStoreTaskDescription: IndexTaskDescription {
+      }
+
+      guard
+        await buildServerManager.toolchain(for: target, language: language)?
+          .canIndexMultipleSwiftFilesInSingleInvocation ?? false
+      else {
+        partitions.append(.singleFile(file: fileInfo, buildSettings: buildSettings))
+        continue
+      }
+
+      // If the build settings contain `-index-unit-output-path`, remove it. We add the index unit output path back in
+      // using an `-output-file-map`. Removing it from the build settings allows us to index multiple Swift files in a
+      // single compiler invocation if they share all build settings and only differ in their `-index-unit-output-path`.
+      let indexUnitOutputPathFromSettings = removeIndexUnitOutputPath(
+        from: &buildSettings,
+        for: fileInfo.mainFile
+      )
+
+      switch (fileInfo.outputPath, indexUnitOutputPathFromSettings) {
+      case (.notSupported, nil):
+        partitions.append(.singleFile(file: fileInfo, buildSettings: buildSettings))
+      case (.notSupported, let indexUnitOutputPathFromSettings?):
+        swiftFileInfosByBuildSettings[buildSettings, default: []].append(
+          (fileInfo.file, indexUnitOutputPathFromSettings)
+        )
+      case (.path(let indexUnitOutputPath), nil):
+        swiftFileInfosByBuildSettings[buildSettings, default: []].append((fileInfo.file, indexUnitOutputPath))
+      case (.path(let indexUnitOutputPath), let indexUnitOutputPathFromSettings?):
+        if indexUnitOutputPathFromSettings != indexUnitOutputPath {
+          logger.error(
+            "Output path reported by BSP server does not match -index-unit-output path in compiler arguments: \(indexUnitOutputPathFromSettings) vs \(indexUnitOutputPath)"
+          )
+        }
+        swiftFileInfosByBuildSettings[buildSettings, default: []].append((fileInfo.file, indexUnitOutputPath))
+      }
+    }
+    for (buildSettings, fileInfos) in swiftFileInfosByBuildSettings {
+      partitions.append(.multipleFiles(filesAndOutputPaths: fileInfos, buildSettings: buildSettings))
+    }
+
+    for partition in partitions {
+      guard let toolchain = await buildServerManager.toolchain(for: target, language: partition.buildSettings.language)
+      else {
         logger.fault(
-          "Unable to determine toolchain to index \(buildSettings.language.description, privacy: .public) files in \(target.forLogging)"
+          "Unable to determine toolchain to index \(partition.buildSettings.language.description, privacy: .public) files in \(target.forLogging)"
         )
         continue
       }
       let startDate = Date()
-      switch buildSettings.language.semanticKind {
+      switch partition.buildSettings.language.semanticKind {
       case .swift:
         do {
           try await updateIndexStore(
-            forSwiftFile: fileInfo.mainFile,
-            buildSettings: buildSettings,
+            forSwiftFilesInPartition: partition,
+            buildSettings: partition.buildSettings,
```

Comment:
Could just pass the partition rather than both the partition + build settings?

Reply:
Good catch. Was an artifact from a previous iteration.